Add an 'extras' property to comments

crazytonyli opened the pull request:
@oguzkocer I tried the "serde flatten" solution you mentioned in our 1:1, and it worked out great. One minor downside is that there will be additional parsing work required for all API endpoints. I tried to use `RawValue` to avoid that, but it does not seem to work with `serde(flatten)` (see https://github.com/serde-rs/json/pull/1252).

WP.com endpoints may return extra information in the responses. We'll need to parse that information to support the app's functionalities. This PR introduces a mechanism to support that. We can add this block of code to the models if needed:

```rust
    #[serde(flatten)]
    #[WpContext(edit, embed, view)]
    pub extras: Option<JsonValue>,
```

crazytonyli committed:
Add an 'extras' property to comments

oguzkocer reviewed (commented):
@crazytonyli I think this is a good start, but it looks like we have quite a few failing tests.

* We could consider marking the field with `#[WpContextualOption]` because there won't always be extra fields. I wonder if that'd also help with de-serializing speed.
* We might want to introduce a new type to hold the extras, instead of directly using `JsonValue`. This might be changed to be a `uniffi::Object` with some helpers in the future and having a type in advance could make things easier for us. It'd also look a bit more consistent when we use it in other types. Note that you probably need to use `#[serde(transparent)]` for this to work, but I don't know if `#[serde(transparent)]` & `#[serde(flatten)]` works together or not.
* The field ordering _might_ matter, as I believe it does for de-serializing `enum`s. That'd be a bit of an issue, because if we accidentally put this field above others, the ones below might always be set to `None`. If this is an issue, we could add a check in `WpContext` proc macro to ensure that if there is an `extra` field, it's at the end. Having a specific type would also help with this.
* If we have a separate type, we could consider moving the `meta` field to it. I am not so sure about this though, so just dropping it as an idea for now.

crazytonyli commented:
> The field ordering might matter

It does not seem so: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=b62029454b6f50395fbb3dfd17a8d4b4.

BTW, I'll leave this PR open until we have an API endpoint that requires this change, like the WP.com /wp/v2/comments (see https://linear.app/a8c/issue/AINFRA-939).

crazytonyli committed:
Parse additional data in WP.com comments API

crazytonyli commented:
@oguzkocer This PR is ready to be reviewed and merged. The `WpComCommentExtension` struct contains data that only exists in the WP.com API.

crazytonyli commented on wp_api/src/comments.rs:
```diff
@@ -528,6 +528,9 @@ pub struct SparseComment {
     pub author_avatar_urls: Option<HashMap<UserAvatarSize, WpResponseString>>,
+    #[serde(flatten)]
+    #[WpContext(edit, embed, view)]
+    pub additional_fields: Option<Arc<AnyJson>>,
```
This field is included in the generated `SparseCommentFields`, which it should not be. I guess we'll have to implement a logic to exclude certain fields. We can either check the `#[serde(flatten)]` attribute (if that's possible), or add a new attribute like `#[WpContextualExcludeFromFields]`. What do you think? @oguzkocer

oguzkocer replied:
I was checking what needed to be done for this and figured I'd quickly implement it, so you don't have to do the same: #833

crazytonyli replied:
Thanks!

I have got this error using the new attribute, do you mind having a look?

```
$ cargo build
   Compiling wp_api v0.1.0 (/Users/tonyli/Projects/wordpress-rs/wp_api)
error[E0599]: no variant or associated item named `AdditionalFields` found for enum `SparseCommentFieldWithEditContext` in the current scope
   --> wp_api/src/comments.rs:495:57
    |
495 | #[derive(Debug, Serialize, Deserialize, uniffi::Record, WpContextual)]
    |                                                         ^^^^^^^^^^^^
    |                                                         |
    |                                                         variant or associated item not found in `SparseCommentFieldWithEditContext`
    |                                                         variant or associated item `AdditionalFields` not found for this enum
    |
    = note: this error originates in the derive macro `WpContextual` (in Nightly builds, run with -Z macro-backtrace for more info)
```

oguzkocer replied:
Fixed in c0cbafddf532a00296b120995df46923c098964d. (after checking with @crazytonyli that it'd be OK for me to push directly to his branch)

oguzkocer commented on wp_api/src/wp_com/endpoint/extensions/comments.rs:
```diff
@@ -0,0 +1,38 @@
+
+#[uniffi::export(with_foreign)]
+pub trait WpComCommentExtensionProvider: Send + Sync {
+    fn parse_extension(&self) -> Result<WpComCommentExtension, UniffiSerializationError>;
```
We probably need to rename this to `parse_wp_com_comment_extension` as otherwise it'll create a conflict if `parse_extension` is reused for another similar trait.

oguzkocer commented on wp_api_integration_tests/tests/test_comments_immut.rs:
```diff
@@ -207,6 +209,73 @@ async fn list_comments_with_edit_context_parse_author_avatar_urls(
+#[tokio::test]
+#[parallel]
+#[ignore]
+async fn wpcom_comment_extension() {
```
We should keep everything related to `wp_com` contained, so that we can move them to `wp_com` crate when the time comes. Can you create a new file for `wp_com` comments tests? See `wp_api_integration_tests/tests/test_wp_com_subscribers_immut.rs` for an example.

oguzkocer commented on wp_api_integration_tests/tests/test_comments_immut.rs:
```diff
@@ -207,6 +209,73 @@ async fn list_comments_with_edit_context_parse_author_avatar_urls(
+#[parallel]
+#[ignore]
+async fn wpcom_comment_extension() {
+    // You'll need to replace the site ID and comment ID with valid ones.
```
Consider using `WpComTestCredentials` for this and adding the field to `wp_com_test_credentials.json-example`.

oguzkocer commented on wp_api_integration_tests/tests/test_comments_immut.rs:
```diff
@@ -207,6 +209,73 @@ async fn list_comments_with_edit_context_parse_author_avatar_urls(
+            middleware_pipeline: Arc::new(WpApiMiddlewarePipeline::default()),
+            app_notifier: Arc::new(EmptyAppNotifier),
+        },
+    );
```
Probably worth adding a helper for this, so we can re-use in other tests. See `wp_api_integration_tests/src/lib.rs` for various helpers. (We need to move the `wp_com` stuff out of that module, but that can be done at a later time)

oguzkocer commented on wp_api_integration_tests/tests/test_comments_immut.rs:
```diff
@@ -207,6 +209,73 @@ async fn list_comments_with_edit_context_parse_author_avatar_urls(
+        .await
+        .assert_response()
+        .data;
+    assert!(comment.additional_fields.parse_extension().is_ok());
```
It'd be best to have a single call per test, both for efficiency and to follow existing conventions.

oguzkocer reviewed (changes requested):
@crazytonyli This is mostly looking good. I've left some suggestions. Let me know what you think!

crazytonyli committed:
Use the new attribute

oguzkocer committed:
Fix WpContextualExcludeFromFields filtering in test helpers

Exclude fields marked with WpContextualExcludeFromFields from integration
test helper generation to prevent referencing non-existent enum variants.

crazytonyli committed:
Rename the parse_extension function

crazytonyli committed:
Move wpcom comments tests

crazytonyli committed:
Breakdown into three test functions for each context

crazytonyli committed:
Format code

crazytonyli commented:
Hey @oguzkocer , I have addressed your comments. Do you mind having another look? Thanks.
